server/mvcc/backend/hooks_test.go

```diff
@@ -0,0 +1,129 @@
+// Copyright 2021 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package backend_test
+
+import (
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"go.etcd.io/etcd/server/v3/mvcc/backend"
+	betesting "go.etcd.io/etcd/server/v3/mvcc/backend/testing"
+)
+
+var (
+	bucket = []byte("bucket")
+	key    = []byte("key")
+)
+
+func TestBackendPreCommitHook(t *testing.T) {
+	be := newTestHooksBackend(t, backend.DefaultBackendConfig())
+
+	tx := be.BatchTx()
+	prepareBuckenAndKey(tx)
+	tx.Commit()
+
+	// Empty commit.
+	tx.Commit()
+
+	write(tx, []byte("foo"), []byte("bar"))
+
+	assert.Equal(t, ">cc", getCommitsKey(t, be), "expected 2 explict commits")
+	tx.Commit()
+	assert.Equal(t, ">ccc", getCommitsKey(t, be), "expected 3 explict commits")
+}
+
+func TestBackendAutoCommitLimitHook(t *testing.T) {
+	cfg := backend.DefaultBackendConfig()
+	cfg.BatchLimit = 3
+	be := newTestHooksBackend(t, cfg)
+
+	tx := be.BatchTx()
+	prepareBuckenAndKey(tx) // writes 2 entries.
+
+	for i := 3; i <= 9; i++ {
+		write(tx, []byte("i"), []byte{byte(i)})
+	}
+
+	assert.Equal(t, ">ccc", getCommitsKey(t, be))
+}
+
+func write(tx backend.BatchTx, k, v []byte) {
+	tx.Lock()
+	defer tx.Unlock()
+	tx.UnsafePut(bucket, k, v)
+}
+
+func TestBackendAutoCommitBatchIntervalHook(t *testing.T) {
+	cfg := backend.DefaultBackendConfig()
+	cfg.BatchInterval = 10 * time.Millisecond
+	be := newTestHooksBackend(t, cfg)
+	tx := be.BatchTx()
+	prepareBuckenAndKey(tx)
+
+	// Edits trigger an auto-commit
+	for getCommitsKey(t, be) != ">c" {
```

Comment:
Could we rather sleep and then assert `getCommitsKey(t, be) != ">c"` or add some timeout here? I'm afraid that this test would hang forever instead of failing when conditions are not met.

Reply:
Done. PTAL.

Reply:
There are 2 analogous places below (`getCommitsKey(t, be) != ">cc"` and `getCommitsKey(t, be) != ">ccc"`). What you think about something like?

```go
func waitUntil(ctx context.Context, f func() bool) error {
	for !f() {
		select {
		case <-ctx.Done():
			return ctx.Err()
		default:
		}
		time.Sleep(10 * time.Millisecond)	
	}
	return nil
}
```

and usage 

```go
err := waitUntil(ctx, func() bool {
	return getCommitsKey(t, be) == ">c"
})
```
?

Reply:
Great suggestion. Fixed.

Reply:
LGTM